plugins/kinetic-dp/fu-kinetic-dp-puma-aux-isp.c

```diff
@@ -0,0 +1,863 @@
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(FuFirmware) img = NULL;
+	g_autoptr(GBytes) app = NULL;
+	g_autoptr(GBytes) isp_drv = NULL;
+	guint8 mca_oui[DPCD_SIZE_IEEE_OUI] = {MCA_OUI_BYTE_0, MCA_OUI_BYTE_1, MCA_OUI_BYTE_2};
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 5);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 5);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 90);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 5);
+
+	priv_puma_aux_isp->isp_processed_size = 0;
+	priv_puma_aux_isp->isp_total_size =
+	    fu_kinetic_dp_firmware_get_arm_app_code_size(firmware_self);
+	g_debug("start Puma AUX-ISP [%s]...",
+		fu_kinetic_dp_aux_isp_get_chip_id_str(dev_info->chip_id));
+	connection = fu_kinetic_dp_connection_new(fu_udev_device_get_fd(FU_UDEV_DEVICE(device)));
+
+	/* write MCA OUI */
+	if (!fu_kinetic_dp_aux_dpcd_write_oui(connection, mca_oui, error)) {
+		goto PUMA_AUX_ISP_END;
+	}
+	fu_progress_step_done(progress);
+
+	/* only load driver if in IROM mode */
+	if (!is_app_mode) {
+		g_debug("loading isp driver because it is in IROM mode");
+		priv_puma_aux_isp->isp_total_size +=
+		    fu_kinetic_dp_firmware_get_isp_drv_size(firmware_self);
+		/* get image of ISP driver */
+		img = fu_firmware_get_image_by_idx(firmware, FU_KT_FW_IMG_IDX_ISP_DRV, error);
+		if (NULL == img)
+			return FALSE;
+		isp_drv = fu_firmware_write(img, error);
+		if (isp_drv == NULL)
+			return FALSE;
+		/* send ISP driver and execute it */
+		payload_data = g_bytes_get_data(isp_drv, &payload_len);
+		if (payload_len > 0)
+			if (!fu_kinetic_dp_puma_aux_isp_send_isp_drv(
+				priv_puma_aux_isp,
+				device,
+				connection,
+				is_app_mode,
+				payload_data,
+				payload_len,
+				fu_progress_get_child(progress),
+				error))
+				goto PUMA_AUX_ISP_END;
+	}
+	fu_progress_step_done(progress);
+
+	/* enable FW update mode */
+	if (!fu_kinetic_dp_puma_aux_isp_enable_fw_update_mode(priv_puma_aux_isp,
+							      firmware_self,
+							      connection,
+							      is_app_mode,
+							      error))
+		goto PUMA_AUX_ISP_END;
+
+	/* get image of App FW */
+	img = fu_firmware_get_image_by_idx(firmware, FU_KT_FW_IMG_IDX_APP_FW, error);
+	if (NULL == img)
+		return FALSE;
+	app = fu_firmware_write(img, error);
+	if (app == NULL)
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* send App FW image */
+	payload_data = g_bytes_get_data(app, &payload_len);
+	g_debug("sending App Firmware starts...");
+	if (!fu_kinetic_dp_puma_aux_isp_send_payload(priv_puma_aux_isp,
+						     connection,
+						     payload_data,
+						     (const guint32)payload_len,
+						     fu_progress_get_child(progress),
+						     (guint32)PUMA_CHUNK_PROCESS_MAX_WAIT,
+						     error,
+						     FALSE)) {
+		g_prefix_error(error, "sending App Firmware payload failed: ");
+		goto PUMA_AUX_ISP_END;
```

Comment:
you need to return FALSE here otherwise the error will be re-used. If the reset always needs to be called even on error use  fu_device_locker_new_full local to this function.

Reply:
fixed